Shodan Search Command Enhancement [CIAC-11976]

mikejrizzo opened the pull request:
<!-- REMINDER: THIS IS A PUBLIC REPO DO NOT POST HERE SECRETS/SENSITIVE DATA -->
## Contributing to Cortex XSOAR Content
Make sure to register your contribution by filling the [contribution registration form](https://forms.gle/XDfxU4E61ZwEESSMA)

**The Pull Request will be reviewed only after the contribution registration form is filled.**

## Status
- [x] In Progress
- [ ] Ready
- [ ] In Hold - (Reason for hold)

## Related Issues
fixes: [CIAC-11976](https://jira-dc.paloaltonetworks.com/browse/CIAC-11976)

## Description
Enhanced the Shodan_v2 integration `search` command to provide additional information:
  - `ntlm` and `ssl > certs` Banner properties, if available
  - Optional Banner properties, if present
  - Added `return_json` command argument to optionally output the entire Shodan API search results including all properties, including those not returned to Context

## Must have
- [x] Tests
- [x] Documentation

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/demisto/content?pullRequest=39525) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/demisto/content?pullRequest=39525) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/demisto/content?pullRequest=39525) it.</sub>

MosheEichler commented on Packs/Shodan/Integrations/Shodan_v2/Shodan_v2.py:
```diff
@@ -209,21 +260,42 @@ def search_command():
                         "Latitude": location["latitude"],
                         "Country": location["country_name"],
                     },
-                    "Timestamp": match.get("timestamp", ""),
-                    "Domains": match.get("domains", []),
-                    "OS": match.get("os", ""),
+                    'Timestamp': match.get('timestamp', ''),
+                    'Domains': match.get('domains', []),
+                    'OS': match.get('os', ''),
+                    'Data': match.get('data', ''),
+                    'Hash': match.get('hash', 0)
```
Please keep all in single or double quotes

mikejrizzo replied:
Changed my single quotes to double quotes to match style used throughout script.

ShirleyDenkberg commented on Packs/Shodan/ReleaseNotes/1_2_3.md:
```diff
@@ -3,4 +3,8 @@
 ##### Shodan v2
 
-- Metadata and documentation improvements.
+- Improved implementation of the **search** command
```
```suggestion
- Improved implementation of the ***search*** command.
```

mikejrizzo replied:
I restored the 1_2_3.md RN from master that had not been merged into my dev branch and bumped my contribution to 1.2.4.  I incorporated your suggested change in that RN.

ShirleyDenkberg commented on Packs/Shodan/ReleaseNotes/1_2_3.md:
```diff
@@ -3,4 +3,8 @@
 
-- Metadata and documentation improvements.
+- Improved implementation of the **search** command
+  - Added useful details from Shodan search results to Context and War Room outputs
+    - Details about SSL/TLS certificates
+    - NTLM properties, often included in results for Remote Desktop Protocol
+  - Added new  *return_json* argument, which returns the complete Shodan search results as JSON
```
```suggestion
  - Added useful details from Shodan search results to Context and War Room outputs.
    - Details about SSL/TLS certificates.
    - NTLM properties, often included in results for Remote Desktop Protocol.
  - Added new  *return_json* argument, which returns the complete Shodan search results as JSON.
```

mikejrizzo replied:
I restored the 1_2_3.md RN from master that had not been merged into my dev branch and bumped my contribution to 1.2.4.  I incorporated your suggested change in that RN.

mikejrizzo commented on Packs/Shodan/Integrations/Shodan_v2/Shodan_v2.py:
```diff
@@ -32,7 +32,48 @@
+    'expired': ('Expired', False),
+    'expires': ('Expires', ''),
+    'issued': ('Issued', ''),
+    'issuer': ('Issuer', {}),
+    'serial': ('Serial', 0),
+    'sig_alg': ('SigAlg', ''),
+    'subject': ('Subject', {}),
```
The YML and README should be correct for this.  The objects in the API response will have one or more keys defined, each of which are mapped separately to their own outputs such as `Shodan.Banner.Ssl.Cert.Subject.CN`, all of which are `string` items per the JSON schema at https://datapedia.shodan.io/banner.schema.json.  Let me know if you don't agree.

mikejrizzo commented on Packs/Shodan/Integrations/Shodan_v2/Shodan_v2.py:
```diff
@@ -177,9 +218,10 @@ def test_module():
+    query = demisto.args()['query']
+    facets = demisto.args().get('facets')
+    page = int(demisto.args().get('page', 1))
+    return_json = demisto.args().get('return_json', 'No')
```
Done.

mikejrizzo commented on Packs/Shodan/Integrations/Shodan_v2/Shodan_v2.py:
```diff
@@ -189,9 +231,18 @@ def search_command():
+    matches = res.get('matches', [])
+
+    # Return the full list of match data as a JSON file if requested
+    if return_json == 'Yes':
```
Done.

MosheEichler commented:
Hi @mikejrizzo, the code looks good!

We're ready for a demo. Please check [this page](https://xsoar.pan.dev/docs/contributing/demo-prep), and let me know when you're available for one over DFIR. 
Feel free also to send me a recording of a demo.

MosheEichler commented:
Thank you very much for your contribution! Unfortunately, your PR review will be slightly delayed because of an Israeli holiday in the upcoming week (the 12th of April - the 19th of April). Thank you in advance for your patience and understanding.

mikejrizzo commented:
Hi Moshe,

Thanks for the review!  I've prepared a video
<https://drive.google.com/file/d/1R8Jhl65P-b_yDjIloZtsKU19B5d1oabB/view?usp=sharing>
demonstrating the modifications I've made.  If there's any questions you
have after watching it let me know and I'd be happy to schedule a call to
demo it live.


Mike Rizzo |  Sr Principal Security Architect | GCIH

Palo Alto Networks  |  3000 Tannery Way  |  Santa Clara, CA 95054  |  USA

Mobile: 616.516.4763  | www.paloaltonetworks.com
<https://www.paloaltonetworks.com/>


<https://www.paloaltonetworks.com/>
<https://www.linkedin.com/company/palo-alto-networks>
<https://www.facebook.com/PaloAltoNetworks/>

The content of this message is the proprietary and confidential property of
Palo Alto Networks, and should be treated as such. If you are not the
intended recipient and have received this message in error, please delete
this message from your computer system and notify me immediately by e-mail.
Any unauthorized use or distribution of the content of this message is
prohibited.



On Thu, Apr 10, 2025 at 6:31 AM Moshe Eichler ***@***.***>
wrote:

> Hi @mikejrizzo
> <https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_mikejrizzo&d=DwMFaQ&c=V9IgWpI5PvzTw83UyHGVSoW3Uc1MFWe5J8PTfkrzVSo&r=UneAo3lfFP6Ug2cmNn9ILrn5Zl-heoEh6h1tBxQf82I&m=sznWIPGxuqLPuTh17a9jTTOov-CZwBG1hdY9YdU-EXh4vzbHdOB6UiLK6gOhSqKk&s=SWPpDs4cq3PXLg90M8R-RJxWR2ypRjpxO_kHfhzipO0&e=>,
> the code looks good!
>
> We're ready for a demo. Please check this page
> <https://xsoar.pan.dev/docs/contributing/demo-prep>, and let me know when
> you're available for one over DFIR.
> Feel free also to send me a recording of a demo.
>
> —
> Reply to this email directly, view it on GitHub
> <https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_demisto_content_pull_39525-23issuecomment-2D2792295825&d=DwMFaQ&c=V9IgWpI5PvzTw83UyHGVSoW3Uc1MFWe5J8PTfkrzVSo&r=UneAo3lfFP6Ug2cmNn9ILrn5Zl-heoEh6h1tBxQf82I&m=sznWIPGxuqLPuTh17a9jTTOov-CZwBG1hdY9YdU-EXh4vzbHdOB6UiLK6gOhSqKk&s=JMmcKJPIhPrVn-T2GeReDSw3dZ_lOiA_DUTGe_Ry5a4&e=>,
> or unsubscribe
> <https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_notifications_unsubscribe-2Dauth_ADC66V3X4K3PNX6KX2HCZWD2YZB7RAVCNFSM6AAAAAB2XNM6G2VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDOOJSGI4TKOBSGU&d=DwMFaQ&c=V9IgWpI5PvzTw83UyHGVSoW3Uc1MFWe5J8PTfkrzVSo&r=UneAo3lfFP6Ug2cmNn9ILrn5Zl-heoEh6h1tBxQf82I&m=sznWIPGxuqLPuTh17a9jTTOov-CZwBG1hdY9YdU-EXh4vzbHdOB6UiLK6gOhSqKk&s=oVVH-IhmxWnXYhDgpVnmQjXNnkXJU9OdMb7Oybefg7A&e=>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
> *MosheEichler* left a comment (demisto/content#39525)
> <https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_demisto_content_pull_39525-23issuecomment-2D2792295825&d=DwMFaQ&c=V9IgWpI5PvzTw83UyHGVSoW3Uc1MFWe5J8PTfkrzVSo&r=UneAo3lfFP6Ug2cmNn9ILrn5Zl-heoEh6h1tBxQf82I&m=sznWIPGxuqLPuTh17a9jTTOov-CZwBG1hdY9YdU-EXh4vzbHdOB6UiLK6gOhSqKk&s=JMmcKJPIhPrVn-T2GeReDSw3dZ_lOiA_DUTGe_Ry5a4&e=>
>
> Hi @mikejrizzo
> <https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_mikejrizzo&d=DwMFaQ&c=V9IgWpI5PvzTw83UyHGVSoW3Uc1MFWe5J8PTfkrzVSo&r=UneAo3lfFP6Ug2cmNn9ILrn5Zl-heoEh6h1tBxQf82I&m=sznWIPGxuqLPuTh17a9jTTOov-CZwBG1hdY9YdU-EXh4vzbHdOB6UiLK6gOhSqKk&s=SWPpDs4cq3PXLg90M8R-RJxWR2ypRjpxO_kHfhzipO0&e=>,
> the code looks good!
>
> We're ready for a demo. Please check this page
> <https://xsoar.pan.dev/docs/contributing/demo-prep>, and let me know when
> you're available for one over DFIR.
> Feel free also to send me a recording of a demo.
>
> —
> Reply to this email directly, view it on GitHub
> <https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_demisto_content_pull_39525-23issuecomment-2D2792295825&d=DwMFaQ&c=V9IgWpI5PvzTw83UyHGVSoW3Uc1MFWe5J8PTfkrzVSo&r=UneAo3lfFP6Ug2cmNn9ILrn5Zl-heoEh6h1tBxQf82I&m=sznWIPGxuqLPuTh17a9jTTOov-CZwBG1hdY9YdU-EXh4vzbHdOB6UiLK6gOhSqKk&s=JMmcKJPIhPrVn-T2GeReDSw3dZ_lOiA_DUTGe_Ry5a4&e=>,
> or unsubscribe
> <https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_notifications_unsubscribe-2Dauth_ADC66V3X4K3PNX6KX2HCZWD2YZB7RAVCNFSM6AAAAAB2XNM6G2VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDOOJSGI4TKOBSGU&d=DwMFaQ&c=V9IgWpI5PvzTw83UyHGVSoW3Uc1MFWe5J8PTfkrzVSo&r=UneAo3lfFP6Ug2cmNn9ILrn5Zl-heoEh6h1tBxQf82I&m=sznWIPGxuqLPuTh17a9jTTOov-CZwBG1hdY9YdU-EXh4vzbHdOB6UiLK6gOhSqKk&s=oVVH-IhmxWnXYhDgpVnmQjXNnkXJU9OdMb7Oybefg7A&e=>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

mikejrizzo commented:
@MosheEichler I've shared a link to a demo video with you directly.  Let me know if you have any questions after viewing it. 

Thanks!

MosheEichler commented:
Hey @mikejrizzo 

The PR CU build fails on 2 errors:

1. [The level of depth for context output path for command: search In the yml should be less or equal to 5 check the following outputs](https://github.com/demisto/content/actions/runs/14364354628/job/40273534413?pr=39525#step:8:364): `[Shodan.Banner.Ssl.Cert.Subject.CN, Shodan.Banner.Ssl.Cert.Issuer.L, Shodan.Banner.Ssl.Cert.Issuer.C, Shodan.Banner.Ssl.Cert.Subject.L, Shodan.Banner.Ssl.Cert.Subject.C, Shodan.Banner.Ssl.Cert.Issuer.O, Shodan.Banner.Ssl.Cert.Issuer.ST, Shodan.Banner.Ssl.Cert.Subject.O, Shodan.Banner.Ssl.Cert.Issuer.OU, Shodan.Banner.Ssl.Cert.Issuer.CN, Shodan.Banner.Ssl.Cert.Subject.ST]`
2. [No release note entry was found for the modelingrule "Shodan Banner Modeling Rule" in the Shodan pack](https://github.com/demisto/content/actions/runs/14364354628/job/40273534413?pr=39525#step:8:370)

Please fix this errors

1. Reduce the context output depth path to be less than 5 (check that the integration still works as expected).
2. Revert the non-relevant changes in the [modelling rule file](https://github.com/demisto/content/pull/39525/files#diff-1c0302d3a0e7824db84c9306dc0b55f64c91aaffc1cce4105ac537a5d6cbeb74)

mikejrizzo commented:
Hi @MosheEichler ,

I've made updates to address the errors you mentioned, and updated the YAML and readme files to reflect the change as well.  I moved the items under `Ssl.Cert` to `SslCert`.  This means the structure of the context entry doesn't completely match the Shodan API response, but now meets the nesting depth requirements.

MosheEichler commented:
Hi @mikejrizzo , the code looks good!

We're ready for a demo. Please check [this page](https://xsoar.pan.dev/docs/contributing/demo-prep), and let me know when you're available for one over DFIR. 
Feel free also to send me a recording of a demo.
